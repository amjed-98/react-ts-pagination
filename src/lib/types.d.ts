import type { CSSProperties } from 'react';

declare global {
  type Color = CSSProperties['color'];
  type BgColor = CSSProperties['backgroundColor'];

  declare type Page = {
    pageNumber: number;
    style: {
      '--color': Color;
      '--background-color': BgColor;
      '--hover-color': Color;
      '--hover-bgColor': BgColor;
    };
    activeClass?: string;
    pageClass?: string;
  };
}
