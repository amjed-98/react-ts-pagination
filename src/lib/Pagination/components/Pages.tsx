import { useMemo, forwardRef, type CSSProperties } from 'react';
import PageNumber from './PageNumber';

type Props = {
  numberOfPages: number;
  currentPageNumber: number;
  handlePageChange: (pageNumber: number) => void;
  pageClass: string;
  activePageClass: string;
  pageStyle?: CSSProperties;
  activePageStyle?: CSSProperties;
  pagesContainerClass: string;
};

const Pages = forwardRef<HTMLDivElement, Props>((props, pagesRef) => {
  const {
    numberOfPages,
    pageClass,
    currentPageNumber,
    activePageClass,
    pageStyle,
    activePageStyle,
    handlePageChange,
    pagesContainerClass,
  } = props;

  const pages = useMemo<Page[]>(() => {
    return Array.from({ length: numberOfPages }, (_, i) => {
      const pageNumber = i + 1;
      const style = pageNumber === currentPageNumber ? activePageStyle : pageStyle;

      return {
        pageNumber,
        style,
        activePageClass,
        pageClass,
      };
    });
  }, [currentPageNumber, numberOfPages, activePageStyle, pageStyle]);

  return (
    <div className={pagesContainerClass} ref={pagesRef}>
      {pages.map((page) => (
        <PageNumber
          key={page.pageNumber}
          page={page}
          currentPageNumber={currentPageNumber}
          handlePageChange={handlePageChange}
        />
      ))}
    </div>
  );
});

export default Pages;
