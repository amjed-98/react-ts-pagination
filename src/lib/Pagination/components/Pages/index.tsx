import { useMemo, forwardRef, type CSSProperties } from 'react';
import Page from '../Page';

export type Props = {
  numberOfPages: number;
  currentPageNumber: number;
  handlePageChange: (pageNumber: number) => void;
  pageClass: string;
  activePageClass: string;
  pagesContainerClass: string;
  pageStyle: CSSProperties;
  activePageStyle: CSSProperties;
  buildPageText: (pageNumber: number) => string | number;
};

const Pages = forwardRef<HTMLDivElement, Props>((props, pagesRef) => {
  const {
    numberOfPages,
    pageClass,
    currentPageNumber,
    activePageClass,
    pageStyle,
    activePageStyle,
    handlePageChange,
    pagesContainerClass,
    buildPageText,
  } = props;

  const pages = useMemo<Page[]>(() => {
    return Array.from({ length: numberOfPages }, (_, i) => {
      const pageNumber = i + 1;
      const style = pageNumber === currentPageNumber ? activePageStyle : pageStyle;

      return {
        pageNumber,
        style,
        activePageClass,
        pageClass,
      };
    });
  }, [currentPageNumber, numberOfPages, activePageStyle, pageStyle, activePageClass, pageClass]);

  return (
    <div className={pagesContainerClass} ref={pagesRef}>
      {pages.map((page) => (
        <Page
          key={page.pageNumber}
          page={page}
          currentPageNumber={currentPageNumber}
          handlePageChange={handlePageChange}
          buildPageText={buildPageText}
        />
      ))}
    </div>
  );
});

export default Pages;
