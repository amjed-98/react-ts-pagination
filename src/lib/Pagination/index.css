#root {
  --color: #a0a7b9;
  --hover-color: white;
  --background-color: transparent;
  --hover-background-color: #8d00d8;
}

.pagination {
  display: flex;
  justify-content: center;
  align-items: baseline;
  padding-block: 4rem;
  margin-top: 1rem;
}

.btn {
  border: none;
  outline: none;
  background-color: transparent;
  cursor: pointer;
  font-size: 1.5rem;
  transition: all 0.1s ease-in-out;
  border-radius: 50%;
  padding: 0.5rem;
  user-select: none;
}

.pages {
  display: flex;
  overflow: hidden;
  max-width: 22rem;
  justify-content: flex-start;
  padding-bottom: 1rem;
  padding-bottom: 1rem;
  gap: 0.8rem;
}

.page {
  font-weight: bold;
  cursor: pointer;
  user-select: none;
  background-color: var(--background-color);
  color: var(--color);
  text-align: center;
  height: 2rem;
  margin: auto 4px;
  display: flex;
  box-sizing: border-box;
  align-items: center;
  justify-content: center;
  letter-spacing: 0.01071em;
  border-radius: 1rem;
  line-height: 1.43;
  font-size: 1rem;
  min-width: 2rem;
}

.active-page {
  background-color: var(--hover-background-color);
  color: var(--hover-color);
}

.page:hover {
  background-color: var(--hover-background-color);
  color: var(--hover-color);
}
