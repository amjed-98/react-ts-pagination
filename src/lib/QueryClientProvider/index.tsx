import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import type { FC, ReactNode } from 'react';

type Props = {
  children: ReactNode | ReactNode[];
};

const queryClient = new QueryClient();

const Provider: FC<Props> = ({ children }) => {
  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

export default Provider;
