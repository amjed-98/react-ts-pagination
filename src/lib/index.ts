export { usePagination, useServerPagination } from './hooks';
export { default as Pagination } from './Pagination';
export { default as Provider } from './QueryClientProvider';
