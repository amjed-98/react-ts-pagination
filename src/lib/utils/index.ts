export { default as scrollToPage } from './scrollToPage';
export { default as getPageRef } from './getPageRef';
export { dispatchEvent, listenFor, unSubscribe } from './event';
export { default as handleMouseWheelScroll } from './handleMouseWheelScroll';
