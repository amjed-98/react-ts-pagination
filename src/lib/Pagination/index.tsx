import type { FC, JSXElementConstructor, CSSProperties } from 'react';
import { useCallback, useEffect, useRef } from 'react';
import {
  dispatchEvent,
  getPageRef,
  listenFor,
  scrollToPage,
  handleMouseWheelScroll,
} from '@/lib/utils';
import Button from './components/Button';
import Pages from './components/Pages';
import './index.css';

export type Props = {
  currentPageNumber: number;
  numberOfPages: number;
  onPageChange?: (pageNumber: number, pageRef: HTMLSpanElement | undefined) => void;
  paginationContainerClass?: string;
  pagesContainerClass?: string;
  nextLabel?: string | JSXElementConstructor<Record<string, never>>;
  prevLabel?: string | JSXElementConstructor<Record<string, never>>;
  nextBtnClass?: string;
  prevBtnClass?: string;
  pageStyle?: CSSProperties;
  pageClass?: string;
  activePageStyle?: CSSProperties;
  activePageClass?: string;
};

const Pagination: FC<Props> = (props) => {
  const {
    currentPageNumber,
    numberOfPages,
    pageStyle,
    activePageStyle,
    pageClass = 'page',
    activePageClass = 'active-page',
    paginationContainerClass = 'pagination',
    pagesContainerClass = 'pages',
    nextBtnClass = 'btn',
    prevBtnClass = 'btn',
    onPageChange,
    prevLabel = '❮',
    nextLabel = '❯',
  } = props;

  const pagesRef = useRef<HTMLDivElement>(null);

  const handlePageChange = useCallback((pageNumber: number): void => {
    /*
      ? if pass  onPageChange handler function then will use it
      ? else will use the default handler that is in usePagination hook
    */
    if (onPageChange) {
      const pageRef = getPageRef(pagesRef, pageNumber);

      onPageChange(pageNumber, pageRef);
      pageRef?.scrollIntoView({ behavior: 'smooth' });
    } else dispatchEvent('pageChange', pageNumber);
  }, []);

  useEffect(
    () =>
      listenFor<number>('pageChange', ({ detail: pageNumber }) =>
        scrollToPage(pagesRef, pageNumber),
      ),
    [],
  );

  return (
    <div className={paginationContainerClass} onWheel={handleMouseWheelScroll} role='pagination'>
      <Button
        Label={prevLabel}
        className={prevBtnClass}
        onClick={handlePageChange.bind(null, currentPageNumber - 1)}
      />

      <Pages
        ref={pagesRef}
        currentPageNumber={currentPageNumber}
        numberOfPages={numberOfPages}
        pageStyle={pageStyle}
        activePageStyle={activePageStyle}
        pageClass={pageClass}
        activePageClass={activePageClass}
        handlePageChange={handlePageChange}
        pagesContainerClass={pagesContainerClass}
      />

      <Button
        Label={nextLabel}
        className={nextBtnClass}
        onClick={handlePageChange.bind(null, currentPageNumber + 1)}
      />
    </div>
  );
};

export default Pagination;
