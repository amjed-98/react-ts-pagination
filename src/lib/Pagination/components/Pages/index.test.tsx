import { render } from '@/lib/test_setup';
import Pages, { type Props } from '.';

const props: Props = {
  numberOfPages: 5,
  pageClass: 'page',
  currentPageNumber: 1,
  activePageClass: 'active',
  handlePageChange: () => {
    console.log('called');
  },
  pagesContainerClass: 'page-container',
};

describe('Page Component', () => {
  const { numberOfPages, pagesContainerClass } = props;
  const { container } = render(<Pages {...props} />);
  const pagesElement = container.children[0];

  it(`should render Page component with ${pagesContainerClass} class`, () => {
    expect(pagesElement).toBeInTheDocument();
    expect(pagesElement.className).toBe(pagesContainerClass);
  });

  it(`should render ${numberOfPages} Pages as the total my pages`, () => {
    expect(pagesElement.children.length).toBe(numberOfPages);
  });

  it('should match the pages snapshot', () => {
    const { baseElement } = render(<Pages {...props} />);

    expect(baseElement).toMatchSnapshot();
  });
});
