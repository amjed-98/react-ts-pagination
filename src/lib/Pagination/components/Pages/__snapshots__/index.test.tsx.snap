// Vitest Snapshot v1

exports[`Page Component > should match the pages snapshot 1`] = `
<body>
  <div>
    <div
      class="page-container"
    >
      <span
        class="page active"
      >
        1
      </span>
      <span
        class="page"
      >
        2
      </span>
      <span
        class="page"
      >
        3
      </span>
      <span
        class="page"
      >
        4
      </span>
      <span
        class="page"
      >
        5
      </span>
    </div>
  </div>
</body>
`;
